File: WebClient SDK/src/default/files/krAddCycleGroupingFileControlExtension.ts
import { reaction } from 'mobx';
import { CycleGrouping } from './cycleGrouping';
import { CycleFilesMode } from './cycleFilesMode';
import { switchFilesVisibility, restoreFilesList } from './cycleGroupingUIHelper';
import { FileControlExtension, IFileControlExtensionContext } from 'tessa/ui/files';
import { ICardModel } from 'tessa/ui/cards';
import { getKrComponentsByCard, KrTypesCache, KrComponents } from 'tessa/workflow';
import { hasNotFlag, createTypedField, DotNetType } from 'tessa/platform';
import { tryGetFromInfo, UIContext } from 'tessa/ui';
import { IStorage } from 'tessa/platform/storage';
import { CardSingletonCache } from 'tessa/cards';

// tslint:disable: triple-equals

/**
 * Управляет появлением виртуальных файлов "версий" при включении и выключении группировки по циклам согласования
 */
export class KrAddCycleGroupingFileControlExtension extends FileControlExtension {
  private _dispose: Function[] = [];

  public initialized(context: IFileControlExtensionContext) {
    const model = context.info['.cardModel'] as ICardModel;
    if (!model || model.inSpecialMode) {
      return;
    }

    const card = model.card;
    if (
      !card ||
      hasNotFlag(getKrComponentsByCard(card, KrTypesCache.instance), KrComponents.Routes)
    ) {
      return;
    }

    let cycleGrouping = context.groupings.find(p => p.name === 'Cycle');

    // Добавляем группировку/фильтр
    if (!cycleGrouping) {
      cycleGrouping = new CycleGrouping('Cycle', '$UI_Controls_FilesControl_GroupingByCycle');
      context.groupings.push(cycleGrouping);
    }

    // Если по умолчанию не выбрана группировка по циклу согласования, то надо убрать виртуальные файлы версий.
    if (!(context.control.selectedGrouping instanceof CycleGrouping)) {
      context.control.removeFiles(file => {
        const cardFile = card.files.find(x => x.rowId == file.id);
        const cycle = tryGetFromInfo<number>(file.model.info, 'KrCycleID');
        return cardFile?.isVirtual !== false && cycle != undefined;
      });
    }

    const sections = card.tryGetSections();
    if (!sections) {
      return;
    }

    const commonInfo = sections.tryGet('KrApprovalCommonInfoVirtual');
    const state: number | null = commonInfo ? commonInfo.fields.tryGet('StateID') : null;

    let currentCycle: number | null = null;
    const approvalHistory = sections.tryGet('KrApprovalHistoryVirtual');
    if (approvalHistory) {
      const rows = approvalHistory.tryGetRows();
      if (rows && rows.length > 0) {
        currentCycle = Math.max(...rows.map(x => x.tryGet('Cycle', 0)));
      }
    }

    this._dispose.push(
      context.control.containerFileAdded.addWithDispose(e => {
        if (
          state != undefined &&
          state !== 0 && // KrState.Draft
          !!e.file.origin &&
          currentCycle != undefined &&
          currentCycle > 0
        ) {
          e.file.info['KrCycleID'] = createTypedField(currentCycle, DotNetType.Int32);
        }
      })!
    );

    this._dispose.push(
      reaction(
        () => context.control.selectedGrouping,
        grouping => {
          if (grouping instanceof CycleGrouping) {
            const currentMode = tryGetFromInfo<CycleFilesMode | null>(
              context.control.info,
              'CycleGroupingMode'
            );
            switchFilesVisibility(
              context.control,
              card,
              currentCycle,
              currentMode == undefined ? CycleFilesMode.ShowAllCycleFiles : currentMode
            );
          } else {
            restoreFilesList(context.control, card);
          }
        }
      )
    );

    let currentCycleMode: CycleFilesMode | null = null;
    let modeFromContext: CycleFilesMode = CycleFilesMode.ShowAllCycleFiles;
    const cardEditor = UIContext.current.cardEditor;
    const groupingModeFromInfo = tryGetFromInfo<IStorage | null>(
      UIContext.current.info,
      'CycleGroupingMode'
    );

    if (
      context.control.name &&
      cardEditor &&
      cardEditor.cardModel &&
      card.id === cardEditor.cardModel.card.id &&
      groupingModeFromInfo &&
      context.control.name in groupingModeFromInfo
    ) {
      modeFromContext = groupingModeFromInfo[context.control.name];
      currentCycleMode = modeFromContext;
    }

    if (currentCycleMode == undefined) {
      const cardModelInfo = model.info;
      const mode = tryGetFromInfo<CycleFilesMode | null>(cardModelInfo, 'CycleGroupingMode');
      if (mode == undefined) {
        const settings = CardSingletonCache.instance.cards.get('KrSettings')!;

        // Проверим, что тип карточки/документа включён в настройки
        // Читеем тип карточки/документа ровно один раз
        let docCardTypeId: guid | undefined = undefined;
        const dciSection = card.sections.tryGet('DocumentCommonInfo');
        if (dciSection) {
          docCardTypeId = dciSection.fields.tryGet('DocTypeID');
        }
        if (docCardTypeId == undefined) {
          docCardTypeId = card.typeId;
        }

        let settingsRowId: guid | undefined = undefined;
        if (
          state != undefined &&
          settings.sections.get('KrSettingsCycleGrouping')!.rows.some(x => {
            const typeId = x.get('TypeID');
            if (typeId === docCardTypeId) {
              settingsRowId = x.get('TypesRowID');
              // Проверим состояния
              if (
                settings.sections
                  .get('KrSettingsCycleGroupingStates')!
                  .rows.filter(y => y.get('TypesRowID') === settingsRowId)
                  .some(y => y.get('StateID') === state)
              ) {
                return true;
              }
            }

            return false;
          })
        ) {
          const row = settings.sections
            .get('KrSettingsCycleGroupingTypes')!
            .rows.find(x => x.rowId === settingsRowId);
          currentCycleMode = row ? row.get('DefaultModeID') : null;
          if (currentCycleMode != undefined && cardModelInfo) {
            cardModelInfo['CycleGroupingMode'] = currentCycleMode;
          }
        }
      }
    }

    if (currentCycleMode != undefined) {
      context.control.info['CycleGroupingMode'] = currentCycleMode;
      if (!context.control.selectedGrouping) {
        context.control.selectedGrouping = cycleGrouping;
      }
    }
  }

  public finalized() {
    for (let dispose of this._dispose) {
      if (dispose) {
        dispose();
      }
    }
    this._dispose.length = 0;
  }
}

File: Source/Extensions/Tessa.Extensions.Default.Client/Files/KrAddCycleGroupingFileControlExtension.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Tessa.Cards;
using Tessa.Cards.Caching;
using Tessa.Extensions.Default.Shared;
using Tessa.Extensions.Default.Shared.Workflow.KrProcess;
using Tessa.Platform.Storage;
using Tessa.UI;
using Tessa.UI.Cards;
using Tessa.UI.Files;

namespace Tessa.Extensions.Default.Client.Files
{
    /// <summary>
    /// Управляет появлением виртуальных файлов "версий" при включении и выключении группировки по циклам согласования
    /// </summary>
    public class KrAddCycleGroupingFileControlExtension :
        FileControlExtension
    {
        #region Fileds

        private readonly IKrTypesCache krTypesCache;
        private readonly ICardCache cardCache;

        #endregion

        #region Constructors

        public KrAddCycleGroupingFileControlExtension(IKrTypesCache krTypesCache, ICardCache cardCache)
        {
            this.cardCache = cardCache;
            this.krTypesCache = krTypesCache;
        }

        #endregion

        #region Base Overrides

        public override async Task Initialized(IFileControlExtensionContext context)
        {
            ICardModel model = context.TryGetCardModel();
            if (model?.InSpecialMode() != false)
            {
                return;
            }

            Card card = model.Card;
            if (card is null
                || (await KrComponentsHelper.GetKrComponentsAsync(card, this.krTypesCache, context.CancellationToken)).HasNot(KrComponents.Routes))
            {
                return;
            }

            var cycleGrouping = context.Groupings.TryGet(CycleFileGroupingNames.Cycle);

            // Добавляем группировку/фильтр
            if (cycleGrouping is null)
            {
                cycleGrouping = new CycleGrouping(CycleFileGroupingNames.Cycle, "$UI_Controls_FilesControl_GroupingByCycle");
                context.Groupings.Add(cycleGrouping);
            }

            // Если по умолчанию не выбрана группировка по циклу согласования, то надо убрать виртуальные файлы версий.
            if (!(context.Control.SelectedGrouping is CycleGrouping))
            {
                for (int i = context.Control.Files.Count - 1; i > -1; i--)
                {
                    var file = context.Control.Files[i];
                    var cardFile = card.Files.FirstOrDefault(x => x.RowID == file.ID);
                    if (cardFile?.IsVirtual != false &&
                        file.Info.TryGetValue(CycleGroupingInfoKeys.CycleIDKey, out object cycleObj) &&
                        cycleObj != null)
                    {
                        context.Control.Files.RemoveAt(i);
                    }
                }
            }

            StringDictionaryStorage<CardSection> sections = card.TryGetSections();
            if (sections is null)
            {
                return;
            }

            int? state = sections.TryGetValue("KrApprovalCommonInfoVirtual", out CardSection commonInfo)
                ? commonInfo.Fields.TryGet<int?>("StateID")
                : null;

            ListStorage<CardRow> rows;
            int? currentCycle = sections.TryGetValue("KrApprovalHistoryVirtual", out CardSection approvalHistory)
                && (rows = approvalHistory.TryGetRows())?.Count > 0
                    ? rows.Max(p => p.TryGet<int>("Cycle")) // TryGet вернёт 0
                    : (int?)null;

            context.Control.ContainerFileAdded += (sender, args) =>
            {
                if (state.HasValue &&
                    state.Value != KrState.Draft &&
                    args.File.Origin != null)
                {
                    if (currentCycle > 0)
                    {
                        args.File.Info[CycleGroupingInfoKeys.CycleIDKey] = currentCycle.Value;
                    }
                }
            };

            context.Control.PropertyChanged += async (sender, args) =>
            {
                if (args.PropertyName == nameof(IFileControl.SelectedGrouping))
                {
                    if (context.Control.SelectedGrouping is CycleGrouping)
                    {
                        CycleFilesMode? currentMode =
                            context.Control.Info.TryGet<CycleFilesMode?>(CycleGroupingInfoKeys.CycleGroupingModeKey);

                        await CycleGroupingUIHelper.SwitchFilesVisibilityAsync(context.Control, card, currentCycle, currentMode ?? CycleFilesMode.ShowAllCycleFiles);
                    }
                    else
                    {
                        await CycleGroupingUIHelper.RestoreFilesListAsync(context.Control, card);
                    }
                }
            };

            CycleFilesMode? currentCycleMode = null;
            CycleFilesMode modeFromContext = default;
            if (context.Control.Name != null &&
                card.ID == UIContext.Current.CardEditor?.CardModel?.Card.ID &&
                UIContext.Current.Info.TryGet<Dictionary<string, CycleFilesMode>>(CycleGroupingInfoKeys
                    .CycleGroupingModeKey)?.TryGetValue(context.Control.Name, out modeFromContext) == true)
            {
                currentCycleMode = modeFromContext;
            }

            if (!currentCycleMode.HasValue)
            {
                var cardModelInfo = context.TryGetCardModel()?.Info;
                var mode = cardModelInfo.TryGet<CycleFilesMode?>(CycleGroupingInfoKeys.CycleGroupingModeKey);
                CardCacheValue<Card> krSettings;
                if (!mode.HasValue
                    && (krSettings = await this.cardCache.Cards.GetAsync("KrSettings", context.CancellationToken)).IsSuccess)
                {
                    Card settings = krSettings.GetValue();

                    // Проверим, что тип карточки/документа включён в настройки
                    // Читеем тип карточки/документа ровно один раз
                    Guid docCardTypeID;
                    if (card.Sections.TryGetValue("DocumentCommonInfo", out var dciSection) &&
                        dciSection.Fields.TryGetValue("DocTypeID", out var docTypeIDobj) &&
                        docTypeIDobj != null)
                    {
                        docCardTypeID = (Guid)docTypeIDobj;
                    }
                    else
                    {
                        docCardTypeID = card.TypeID;
                    }

                    Guid? settingsRowID = null;
                    if (state.HasValue &&
                        settings.Sections["KrSettingsCycleGrouping"]
                            .Rows.Any(p =>
                            {
                                var typeID = p.Fields.Get<Guid>("TypeID");

                                if (typeID == docCardTypeID)
                                {
                                    settingsRowID = p.Fields.Get<Guid>("TypesRowID");

                                    // Проверим состояния
                                    if (settings.Sections["KrSettingsCycleGroupingStates"]
                                        .Rows.Where(q => q.Fields.Get<Guid>("TypesRowID") == settingsRowID)
                                        .Any(q => q.Fields.Get<int>("StateID") == state.Value))
                                    {
                                        return true;
                                    }
                                }

                                return false;
                            }))
                    {
                        currentCycleMode = (CycleFilesMode?) settings.Sections["KrSettingsCycleGroupingTypes"].Rows.FirstOrDefault(p => p.RowID == settingsRowID)
                            ?.TryGet<int?>("DefaultModeID");

                        if (currentCycleMode.HasValue && cardModelInfo != null)
                        {
                            cardModelInfo[CycleGroupingInfoKeys.CycleGroupingModeKey] = currentCycleMode.Value;
                        }
                    }
                }
            }

            if (currentCycleMode.HasValue)
            {
                context.Control.Info[CycleGroupingInfoKeys.CycleGroupingModeKey] = currentCycleMode;

                if (context.Control.SelectedGrouping == null)
                {
                    await context.Control.SelectGroupingAsync(cycleGrouping, context.CancellationToken);
                }
            }
        }

        #endregion
    }
}
